sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DiagnosticsProvider.java

```diff
@@ -0,0 +1,1345 @@
+            operationType,
+            resourceType,
+            null,
+            CosmosBatchResponse::getStatusCode,
+            (r) -> null,
+            CosmosBatchResponse::getRequestCharge,
+            CosmosBatchResponse::getDiagnostics,
+            thresholds);
+    }
+
+    public <T> Mono<CosmosItemResponse<T>> traceEnabledCosmosItemResponsePublisher(
+       Mono<CosmosItemResponse<T>> resultPublisher,
+       Context context,
+       String spanName,
+       String containerId,
+       String databaseId,
+       CosmosAsyncClient client,
+       ConsistencyLevel consistencyLevel,
+       OperationType operationType,
+       ResourceType resourceType,
+       CosmosDiagnosticsThresholds thresholds) {
+
+        checkNotNull(client, "Argument 'client' must not be null.");
+
+        String accountName = clientAccessor.getAccountTagValue(client);
+
+        return publisherWithDiagnostics(
+            resultPublisher,
+            context,
+            spanName,
+            containerId,
+            databaseId,
+            accountName,
+            client,
+            consistencyLevel,
+            operationType,
+            resourceType,
+            null,
+            CosmosItemResponse::getStatusCode,
+            (r) -> null,
+            CosmosItemResponse::getRequestCharge,
+            CosmosItemResponse::getDiagnostics,
+            thresholds);
+    }
+
+    /**
+     * Runs given {@code Flux<T>} publisher in the scope of trace context passed in using
+     * {@link DiagnosticsProvider#setContextInReactor(Context)} in {@code contextWrite}
+     * Populates active trace context on Reactor's hot path. Reactor's instrumentation for OpenTelemetry
+     * (or other hypothetical solution) will take care of the cold path.
+     *
+     * @param publisher publisher to run.
+     * @return wrapped publisher.
+     */
+    public <T> Flux<T> runUnderSpanInContext(Flux<T> publisher) {
+        return propagatingFlux
+            .flatMap(ignored -> publisher);
+    }
+
+    private <T> Mono<T> diagnosticsEnabledPublisher(
+      CosmosDiagnosticsContext cosmosCtx,
+      Mono<T> resultPublisher,
+      Context context,
+      String spanName,
+      Function<T, Integer> statusCodeFunc,
+      Function<T, Integer> actualItemCountFunc,
+      Function<T, Double> requestChargeFunc,
+      Function<T, CosmosDiagnostics> diagnosticsFunc
+    ) {
+
+        if (!isEnabled()) {
+            return resultPublisher;
+        }
+
+        Optional<Object> callDepth = context.getData(COSMOS_CALL_DEPTH);
+        final boolean isNestedCall = callDepth.isPresent();
+        if (isNestedCall) {
+            return resultPublisher;
+        }
+
+        // propagatingMono ensures active span is propagated to the `resultPublisher`
+        // subscription and hot path. OpenTelemetry reactor's instrumentation will
+        // propagate it on the cold path.
+        return propagatingMono
+            .flatMap(ignored -> resultPublisher)
```

Comment:
I think this is just pushing the tracing to a different thread - for example off of the I/O thread - but I agree we should investigate the right option - my preference would be to do what the COSMOS.SWITCH_OFF_IO_THREAD_FOR_RESPONSE is doing and see whether we can keep tracing and diagnostics handling on the hot path. Otherwise you just add more work for context swiches - it is not like we have any control/threshold where we would skip diagnostics processing under high load etc. - but needs perf evaluation - I am tracking this in github issue https://github.com/Azure/azure-sdk-for-java/issues/34299